game/hud/src/services/session/layout.ts

```
@@ -109,18 +121,80 @@ const initialState = {
-  var storedState = JSON.parse(localStorage.getItem(localStorageKey)) as LayoutState;
+interface Size {
+  width: number;
+  height: number;
+}
+
+const px2pcnt = (pos: Position, screen: Size) : Position => {
+  // converts pixel co-ordinates into % of display size
+  return {
+    x: pos.x / screen.width,
+    y: pos.y / screen.height,
+    width: pos.width,           // don't scale width/height
+    height: pos.height,
+    scale: 1
+  };
+};
+
+const pcnt2px = (pos: Position, screen: Size) : Position => {
+  // converts % co-ordinates into pixels for current display
+  return {
+    x: screen.width * pos.x,
+    y: screen.height * pos.y,
+    width: pos.width,           // don't scale width/height
+    height: pos.height,
+    scale: 1
+  }
+};
+
+const forceOnScreen = (pos: Position, screen: Size) : Position => {
+  if (pos.x < 0) pos.x = 0;
+  if (pos.y < 0) pos.y = 0;
+  if (pos.x + pos.width > screen.width) pos.x = screen.width - pos.width;
+  if (pos.y + pos.height > screen.height) pos.y = screen.height - pos.height;
+  if (pos.x < 0) { pos.x = 0; pos.width = screen.width; }
+  if (pos.y < 0) { pos.y = 0; pos.height = screen.height; }
+  return pos;
+};
+
+const getInitialState = (): any => {
+  const storedState: LayoutState = loadState();
   if (storedState) {
     storedState.locked = initialState.locked;
     return storedState;
   }
   return initialState;
 }
 
+const loadState = () : LayoutState => {
+  const screen: Size = { width: window.innerWidth, height: window.innerHeight };
+  const state: LayoutState = JSON.parse(localStorage.getItem(localStorageKey)) as LayoutState;
+  if (state) {
+    if ((state.version|0) >= MIN_STATE_VERSION_PERCENT) {
+      for (let key in state.widgets) {
+        state.widgets[key] = forceOnScreen(pcnt2px(state.widgets[key], screen), screen);
+      }
+    }
+    return state;
+  }
+};
+
+const saveState = (state: LayoutState) => {
+  const screen: Size = { width: window.innerWidth, height: window.innerHeight };
```

This function can make no assumptions about who/what is calling it. It should validate and fail in an expected manner if the required input is not provided.